test/extended/ha/ha.go

```diff
@@ -0,0 +1,258 @@
+		go func() {
+			io.Copy(g.GinkgoWriter, stderr)
+		}()
+
+		// log stdout and look for exited state
+		go func() {
+			bufReader := bufio.NewReader(stdout)
+			for {
+				line, err := bufReader.ReadString('\n')
+				if err != nil {
+					ec <- err
+					return
+				}
+				g.GinkgoWriter.Write([]byte(fmt.Sprintf("`oc get --watch` output: %s", line)))
+				submatch := reExitCode.FindStringSubmatch(line)
+				if len(submatch) > 1 {
+					if submatch[1] != "0" {
+						ec <- fmt.Errorf("Pod terminated with unexpected exit code (%s)", submatch[1])
+					} else {
+						ec <- nil
+					}
+					return
+				}
+			}
+		}()
+
+		select {
+		case err := <-ec:
+			if err != nil && err != io.EOF {
+				o.Expect(err).NotTo(o.HaveOccurred())
+			}
+			if err == nil {
+				cmd.Process.Kill()
+				cmd.Wait()
+			} else {
+				o.Expect(cmd.Wait()).NotTo(o.HaveOccurred())
+			}
+		case <-time.After(start.Add(timeout).Sub(time.Now())):
+			g.Fail(fmt.Sprintf("Timeout (%s) occurred while waiting for pod's exit", timeout.String()))
+		}
+	}
+
+	StartPod := func() {
+		g.By(fmt.Sprintf("\n HA TEST:  Creating a pod from %s", podConfig))
+		err := oc.Run("create").Args("-f", podConfig).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("\n HA TEST:  Wait for pod's exit")
+		WaitForPod("helloworld-pod", time.Second*50)
+
+		g.By("\n HA TEST:  Check pod's output")
+		out, err := oc.Run("logs").Args("--interactive=false", "helloworld-pod").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(strings.Contains(strings.ToLower(out), "hello from docker"))
+	}
+
+	// EnsureOneInstanceActive releases any inactive controllers instances and
+	// makes sure that one is active.
+	EnsureOneInstanceActive := func() {
+		l := m.Len()
+		switch {
+		case l > 1:
+			g.By("\n HA TEST:  Releasing inactive controllers")
+			m.ReleaseControllers(m.GetInactive()...)
+		case l == 0:
+			g.By("\n HA TEST:  Launching initial controllers instance")
+			// Speed up controllers election by deleting current lease
+			m.DeleteLease()
+			_, err := m.StartNewInstance()
+			o.Expect(err).NotTo(o.HaveOccurred())
+			o.Expect(m.Len()).To(o.Equal(1))
+			_, _, err = m.WaitForActive(time.Second)
+			o.Expect(err).NotTo(o.HaveOccurred())
+		}
+		o.Expect(m.Len()).To(o.Equal(1))
+	}
+
+	TerminateActive := func() *exutil.Controllers {
+		g.By("\n HA TEST:  Terminating active controllers instance")
+		ctrls, err := m.GetActive()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		m.ReleaseControllers(ctrls)
+		o.Expect(ctrls.Exited()).To(o.BeTrue())
+		return ctrls
+	}
+
+	g.JustBeforeEach(func() {
+		oc.SetOutputDir(exutil.TestContext.OutputDir)
+
+		o.Expect(m.Len()).To(o.Equal(1))
+		g.By("\n HA TEST:  waiting for deployer service account")
+		err := exutil.WaitForDeployerAccount(oc.KubeREST().ServiceAccounts(oc.Namespace()))
+		o.Expect(err).NotTo(o.HaveOccurred())
+	})
+
+	g.AfterEach(func() {
+		EnsureOneInstanceActive()
+	})
+
+	g.Describe("\n HA TEST:  Delete an active lease", func() {
```

Comment:
On the messages, I wanted an easier way to be able to filter for the messages from the forcepull test if I ran the whole set of tests (I did not find the ginkgo facilities sufficient).  As you can see, not everyone agrees with my opinion on this ;-).

On BeforeSuite, according to the doc at http://onsi.github.io/ginkgo/#global-setup-and-teardown-beforesuite-and-aftersuite,
it only runs "once before the entire test suite".  And that was my observation when developing.  Not sure what @mfojtik means by "runs for every test".

I do see that the focus string for forcepull was changed to "builds: serial:" since then.  I don't know yet if that has any bearing on the forcepull BeforeSuite behavior (I had not tried extended tests since their latest refactoring and I'm still trying to get them to run as of this writing).